### ISSUE TITLE
Severity and description of many labels are not getting stored in the database

### ISSUE THREAD

**Issue by manaswis** [ProjectSidewalk/SidewalkWebpage#937]
During the interview today, the participant added lots of labels, each one marked with severity and description filled in. I don't see the severity and description associated with the label in the database. 

And for a few labels, the selected severity is there but the description is not.

![image](https://user-images.githubusercontent.com/2873216/28950518-86d28014-7892-11e7-8f56-3b317f47ad67.png)

All of these labels had severity and description.

This needs to be fixed. I am looking into the log data now if the logs are present or not.

**Comment by jonfroehlich** [comment]
Interesting. I thought this was fixed back in 2016 (I had reported it to Kotaro and he thought he had fixed it then). Collecting severity labels and description text is a core part of the system and this bug dramatically decreases the value of our data, the kinds of analyses I was hoping to write about in our paper, and the kinds of tools we can build (e.g., limits the way we can compute AccessScores). In all, I'm quite disappointed and frustrated.

We need to: (i) fix this problem and (ii) determine how widespread it is. For the latter, perhaps we can use the logging system and compare logged events to our data (of course, this only works for data collected starting in late 2016 as the comprehensive logging system didn't exist until then).

@misaugstad, this directly affects our ability to collect ground truth so this should be a priority to investigate and fix.

**Comment by manaswis** [comment]
### Severity Values
I looked at the log data. There is an event called `ContextMenu_RadioChange` that logs the severity value in the note field in the `audit_task_interaction` table. So all the severities of the placed labels are there. The participant had applied 25 labels with severity selected, only 10 of them were saved in the database. However, those 25 labels's `ContextMenu_RadioChange` event has those severity labels. See below:

![image](https://user-images.githubusercontent.com/2873216/28968872-1639adec-78f0-11e7-9527-0300552664bc.png)

The format of the note is `LabelType:<LabeType>,RadioValue:<SeverityValue>`. Looking at the history of commits to the file (first one going back to August 2016), this type of logging has been there. However, I also found some old data from April 2016 that has the format `Radio:<SeverityValue>`.

Another point to note: I looked into the database, the first record in the audit_interaction table with the format `LabelType: label.getProperty("labelType"), RadioValue: severity`, it shows Dec 22 2016, not sure why. Because the logging has been since August 2016 (based on commit history of the file).

### Description Text
I found another log event called `ContextMenu_TextBoxFocus` and the number of instances of this event coincides with the number of labels that had description filled in, 25 for this user for the audit_task `17109`. But there isn't a log where the description text is stored. So I think that data is lost.

Here is the related code for the description text ([ContextMenu.js Line 113](https://github.com/ProjectSidewalk/SidewalkWebpage/blob/83cd409bde3bf39750e11fa9951f9de69eeb68d0/public/javascripts/SVLabel/src/SVLabel/canvas/ContextMenu.js#L113)):

```
    function handleDescriptionTextBoxChange(e) {
        var description = $(this).val(),
            label = getTargetLabel();
        if (label) {
            label.setProperty('description', description);
        }
-----------> SHOULD LOG HERE <-----------------
    }

    function handleDescriptionTextBoxBlur() {
        svl.tracker.push('ContextMenu_TextBoxBlur');
        svl.ribbon.enableModeSwitch();
        svl.keyboard.setStatus('focusOnTextField', false);
    }

    function handleDescriptionTextBoxFocus() {
        svl.tracker.push('ContextMenu_TextBoxFocus');
        svl.ribbon.disableModeSwitch();
        svl.keyboard.setStatus('focusOnTextField', true);
    }
```
Ideally, `handleDescriptionTextBoxChange` function should log the description, which it isn't doing. The rest of the functions are logging.

**Comment by jonfroehlich** [comment]
Thanks for this analysis and your report. Some questions/comments:

- If I understand correctly, we can recover the severity labels by looking at the log tables but *not* the description text. You are recommending that we update the 'handleDescriptionTextBoxChange' function so that we do indeed log the entered description text in the future. If so, please make this a separate Issue in github.
- I'd like to see something like: a breakdown of the total number of labels we would expect to have ratings given logged events and the total number of labels that actually have ratings.
- After the above is down, I'd like to discuss if it's worth it to write a utility method that takes the log event data and uses this to fix errantly blank severity ratings. My gut is to just suggest that we do this outright but I want to see the data first.
- Your analysis focused on analyzing the log file to see if the data is recoverable but not in determining *why* we are losing data in the first place. Who is looking into this? And is it related to using the keyboard vs. mouse or something more nuanced and nefarious.

**Comment by manaswis** [comment]
> - If I understand correctly, we can recover the severity labels by looking at the log tables but *not* the description text. You are recommending that we update the 'handleDescriptionTextBoxChange' function so that we do indeed log the entered description text in the future. If so, please make this a separate Issue in github.

That's correct. I will make one.

> - I'd like to see something like: a breakdown of the total number of labels we would expect to have ratings given logged events and the total number of labels that actually have ratings.
- After the above is down, I'd like to discuss if it's worth it to write a utility method that takes the log event data and uses this to fix errantly blank severity ratings. My gut is to just suggest that we do this outright but I want to see the data first.

Sounds good.

> - Your analysis focused on analyzing the log file to see if the data is recoverable but not in determining *why* we are losing data in the first place. Who is looking into this? And is it related to using the keyboard vs. mouse or something more nuanced and nefarious. 

As noted on Slack, I haven't looked into why the data isn't getting saved yet. Yes, we'd have to see if it has to do with keyboard vs mouse or something else. No one has taken this issue yet. I suggest @sbower213  or @adash12 or @misaugstad since all of them have looked into logging events. I can also work on this but I have to conduct interviews tomorrow (and maybe on Sunday - scheduling is in progress) and the coming week. So won't make progress on this when I am doing that.

**Comment by jonfroehlich** [comment]
@misaugstad can't do this in the shortterm. He has to work on the codebook and prepare for a session with a mobility impaired person today.

Perhaps @adash12 or @sbower213 can do it then?

**Comment by misaugstad** [comment]
@adash12 only works Mon-Wed, so it looks like @sbower213 remains!

**Comment by jonfroehlich** [comment]
I believe @adash12 was here last Friday...

**Comment by sbower213** [comment]
This is recap for all of us who just met:

First and foremost, if the threshold of 200 interactions (at which point we send data to the server) is hit before a user has entered their description, the description isn't stored. The same is true if a user places a label but enters a rating after the 200 interaction mark is sent. 

It is very easy to hit this 200 interaction limit just by moving your mouse to select an issue from the top bar and putting a label onto the map. You can even hit it just while moving your mouse to select a severity rating or enter a description for your label. The reason for this is the extremely high number of interactions associated with mousemove events. 

I will work on creating functionality to update labels after they have been initially placed and stored so that we can reflect in our table any changes that happen to labels after they are first submitted.

**Comment by manaswis** [comment]
## Damage Analysis

Here is the full damage analysis report. 
**NOTE**: The numbers are based on raw counts of the log events, so they aren't very accurate but gives us a rough estimate.

### Severity Labels
Total labels with severity saved in the database: 4215 
Total labels: 9704
**~43% of labels have severity stored**

The maximum number of labels that don’t have severity stored: 5489 (57%)
Raw count of RadioChange events: 6567
Removed duplicates, total: 5542

Total labels expected to have severity: 5542 (5542/9704 = 57%)
**Missing from the database:** 5542 - 4215 = **1327 = ~14%** (1327/9704)

### Description Text

Total labels with description saved in the database (since Dec 2016): 3018
Total labels since Dec 2016: 21122
**~14% of labels have description text stored**

The maximum number of labels that don’t have description text stored in the database: 18104 (86%)
Raw count of TextBoxFocus events: 5864
Removed duplicates, total: 5028

Total labels expected to have description: 5028 (5028/21122 = ~24%)
**Missing from the database:** 5028 - 3018 = **2010 = ~10%** (2010/21122 )

### Temporary Labels
Total labels with temporary label checkbox set in the database (since April 2016): 692
Total labels since April 2016: 7020
**~10% of labels have temporary labels stored**

The maximum number of labels that don’t have temporary label checkbox set in the database: 6328 (90%)
Raw count of CheckboxChange events: 1063
Removed checked: false, total: 1063 - 39 = 1024

Total labels expected to have temp checkbox set: 1024 (1024/7020 = ~15%)
**Missing from the database:** 1024 - 692 = **332 = ~5%** (332/7020)

### Deleted Labels
Total labels in the database (for the audits where the Click_LabelDelete was logged): 14701
Raw count of Click_LabelDelete events: 2127
Removed duplicates, total: 1756

**Total labels expected to be deleted: 1756  = ~12%** (1756/14701)

**Comment by misaugstad** [comment]
For severity ratings analysis, why are you using a sample of only 10k labels when we have over 80k in the database now?

Also, could you check the difference between pre and post increase on logging? Since we had talked about there possibly being way fewer missing severities pre-logging?

**Comment by manaswis** [comment]
I used the sample where the we had data with logs that are of the form: `LabelType:<LabeType>,RadioValue:<SeverityValue>`. The older logs are of the form `RadioValue:<SeverityValue>`. So we don't have much information about the labelType.

> could you check the difference between pre and post increase on logging? Since we had talked about there possibly being way fewer missing severities pre-logging?

Sure. Will look into that.

**Comment by manaswis** [comment]
Here are the numbers for the data pre-comprehensive logging with the log format of `RadioValue:<SeverityValue>`:

Total labels with severity saved in the database: 29391
Total labels: 43671
**~67.3% of labels have severity store**d

Raw count of RadioChange events: 33159
Didn’t use the label count value after removing the duplicates as too many records were getting removed and the count was becoming less than the actual label count in the database
— After removing duplicates: 24016
— could be because we don’t have any information about the label type in the logs

Total labels expected to have severity: 33159 (33159/43671 = ~62%)
**Missing from the database:** 33159 - 29391 **= 3768 = 8.6%** (3768/43671)

**Comment by jonfroehlich** [comment]
Thanks for this continued analysis.

> NOTE: Interesting thing to note here is pre-comprehensive logging and post-comprehensive logging counts don't add up to the total number of labels currently present in the database. I will look into this more.

It may be that in the very beginning, we had very more limited logging.

Could you update your report to have full comprehensive numbers for all labels in the database (including for severity, description text, temporary, and deletes)? This is what we will want to report in the paper. If it's around 10-15% then I think we can likely still provide a full analyses of these things, as we originally planned. We will have a footnote or short explanation about the lost data. The good thing is that there really isn't a systemic bias to how the data was lost--so I don't think the distribution of severity rankins, etc. should really change.

**Comment by manaswis** [comment]
Closed via PR #945
